### ISSUE TITLE
master_test.go:484: unexpected creation success. The following IPs exist:

### ISSUE THREAD

**Issue by deads2k** [kubernetes/kubernetes#36814]
```
master_test.go:484: unexpected creation success. The following IPs exist: []string{"192.168.0.1", "192.168.0.1", "192.168.0.6", "192.168.0.5", "192.168.0.3", "192.168.0.4", "192.168.0.2"}. It should only be possible to allocate 2 IP addresses in this cluster.
```

Seen in https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/36774/pull-kubernetes-unit/5119/ on https://github.com/kubernetes/kubernetes/pull/36774#issuecomment-260636365 .

@wojtek-t possibly related to the changes we just merged?

**Comment by wojtek-t** [comment]
Hmm - this is the same problem that I was fixing. I don't think my PR is a victim here - I think it just didn't fix all problems...
Though I don't really understand why we can end up with two services having the same IP address now.

@lavalamp

**Comment by wojtek-t** [comment]
I spent some time looking into this one, and to be honest I have no idea what this could happen after recent changes from https://github.com/kubernetes/kubernetes/pull/36581

I added few logs in my local branch to confirm that I'm not missing anything and now indeed it looks like it should:
- creating "kubernetes" service correctly allocates "serviceIP" in etcd (this is a single call) [in fact allocation in etcd happens before service object is created]
- test waits until "kubernetes" service is created
- then creating subsequent services also goes through the path that is allocating serviceIPs

The only thing that could potentially break "serviceIP" range is the "Repair" that would overwrite "kubernetes" service IP as non-used. But it cannot start before "kubernetes" service is created, so that's pretty much impossible.

I don't even know what logs we can add to facilitate future debugging...


On the other hand it seems to me that there is a race in "Repair" object:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/controller/repair.go#L108
because "services" here may not contain serviceIP that was already allocated for the service we are currently creating (in service REST we are first allocating the IP in etcd and only then create a service in etcd).
But that race couldn't be the source if the failure you observed...

**Comment by wojtek-t** [comment]
Though looking into etcd logs
```
$ cat etcd.log | grep "Put] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceip"
2016-11-15 05:04:36.730909 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevExist=false remote:127.0.0.1:53388
2016-11-15 05:04:36.754907 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=974&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:36.770802 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=983&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22AQ%3D%3D%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:37.088044 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=990&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:37.092910 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=997&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22AQ%3D%3D%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:37.105050 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=999&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22IQ%3D%3D%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:37.110470 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=1001&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22MQ%3D%3D%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:37.115626 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=1003&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22NQ%3D%3D%22%7D%0A remote:127.0.0.1:53388
2016-11-15 05:04:37.119951 D | v2http: [PUT] /v2/keys/c1da0028-501b-4a06-a75c-adb22c1e95c2/ranges/serviceips?prevIndex=1005&prevValue=%7B%22kind%22%3A%22RangeAllocation%22%2C%22apiVersion%22%3A%22v1%22%2C%22metadata%22%3A%7B%22creationTimestamp%22%3Anull%7D%2C%22range%22%3A%22192.168.0.0%2F29%22%2C%22data%22%3A%22PQ%3D%3D%22%7D%0A remote:127.0.0.1:53388
```

and at the timestamps, it really seems that those PUTs are from correspondingly:
```
[initialization - empty Range]
"kubernetes" service
Repair
svc-0
svc-1
svc-2
svc-3
svc-4
svc-8
```

there is longer pause between Repair and svc-0 and this is coming from:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/master/master_test.go#L453

So it really seems that Repair may be doing something strange, but I have no idea how...

**Comment by dims** [comment]
@wojtek-t : throwing in the milestone v1.5

**Comment by wojtek-t** [comment]
Closing in favor of #37040
